// node-red-contrib-home-assistant-websocket/dist/nodes/switch/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const joi_1 = __importDefault(require("joi"));
const helpers_1 = require("../../common/controllers/helpers");
const Events_1 = __importDefault(require("../../common/events/Events"));
const InputService_1 = __importDefault(require("../../common/services/InputService"));
const globals_1 = require("../../globals");
const migrate_1 = require("../../helpers/migrate");
const node_1 = require("../../helpers/node");
const homeAssistant_1 = require("../../homeAssistant");
const SwitchController_1 = __importDefault(require("./SwitchController"));
const SwitchStatus_1 = __importDefault(require("./SwitchStatus"));
const inputs = {
    enable: {
        messageProp: 'enable',
    },
};
const inputSchema = joi_1.default.object();
function switchNode(config) {
    globals_1.RED.nodes.createNode(this, config);
    this.config = (0, migrate_1.migrate)(config);
    const { entityConfigNode, serverConfigNode } = (0, node_1.getConfigNodes)(this);
    const homeAssistant = (0, homeAssistant_1.getHomeAssistant)(serverConfigNode);
    const entityConfigEvents = new Events_1.default({
        node: this,
        emitter: entityConfigNode,
    });
    const status = new SwitchStatus_1.default({
        config: serverConfigNode.config,
        entityConfigEvents,
        entityConfigNode,
        node: this,
    });
    const inputService = new InputService_1.default({
        inputs,
        nodeConfig: this.config,
        schema: inputSchema,
    });
    entityConfigNode.integration.setStatus(status);
    const controller = new SwitchController_1.default({
        entityConfigEvents,
        entityConfigNode,
        inputService,
        node: this,
        status,
        ...(0, helpers_1.createControllerDependencies)(this, homeAssistant),
    });
    entityConfigEvents.addListeners(controller, [
        [homeAssistant_1.HaEvent.AutomationTriggered, controller.onTrigger],
        [homeAssistant_1.HaEvent.StateChanged, controller.onStateChange],
    ]);
}
exports.default = switchNode;
