// node-red-contrib-home-assistant-websocket/dist/nodes/webhook/WebhookController.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const ExposeAsMixin_1 = __importDefault(require("../../common/controllers/ExposeAsMixin"));
const OutputController_1 = __importDefault(require("../../common/controllers/OutputController"));
const ExposeAsController = (0, ExposeAsMixin_1.default)((OutputController_1.default));
class WebhookController extends ExposeAsController {
    async onReceivedMessage(data) {
        if (!this.isEnabled)
            return;
        const message = {};
        await this.setCustomOutputs(this.node.config.outputProperties, message, {
            config: this.node.config,
            data: data.payload,
            headers: data.headers,
            params: data.params,
        });
        this.status.setSuccess('home-assistant.status.received');
        this.node.send(message);
    }
}
exports.default = WebhookController;
