// node-red-contrib-home-assistant-websocket/dist/nodes/trigger-state/const.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ComparatorPropertyType = exports.MessageType = exports.PropertyType = exports.TargetType = exports.ENABLE = exports.DISABLE = void 0;
exports.DISABLE = 'disable';
exports.ENABLE = 'enable';
var TargetType;
(function (TargetType) {
    TargetType["ThisEntity"] = "this_entity";
    TargetType["EntityId"] = "entity_id";
})(TargetType || (exports.TargetType = TargetType = {}));
var PropertyType;
(function (PropertyType) {
    PropertyType["CurrentState"] = "current_state";
    PropertyType["PreviousState"] = "previous_state";
    PropertyType["Property"] = "property";
})(PropertyType || (exports.PropertyType = PropertyType = {}));
var MessageType;
(function (MessageType) {
    MessageType["Default"] = "default";
    MessageType["Custom"] = "custom";
    MessageType["Payload"] = "payload";
})(MessageType || (exports.MessageType = MessageType = {}));
var ComparatorPropertyType;
(function (ComparatorPropertyType) {
    ComparatorPropertyType["Always"] = "always";
    ComparatorPropertyType["CurrentState"] = "current_state";
    ComparatorPropertyType["PreviousState"] = "previous_state";
    ComparatorPropertyType["Property"] = "property";
})(ComparatorPropertyType || (exports.ComparatorPropertyType = ComparatorPropertyType = {}));
