// node-red-contrib-home-assistant-websocket/dist/nodes/tag/migrations.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = [
    {
        version: 0,
        up: (schema) => {
            const newSchema = {
                ...schema,
                version: 0,
            };
            return newSchema;
        },
    },
    {
        version: 1,
        up: (schema) => {
            const newSchema = {
                ...schema,
                version: 1,
                outputProperties: [
                    {
                        property: 'payload',
                        propertyType: 'msg',
                        value: '',
                        valueType: 'eventData',
                    },
                    {
                        property: 'topic',
                        propertyType: 'msg',
                        value: '',
                        valueType: 'triggerId',
                    },
                ],
            };
            return newSchema;
        },
    },
    {
        version: 2,
        up: (schema) => {
            var _a;
            const newSchema = {
                ...schema,
                version: 2,
                exposeAsEntityConfig: (_a = schema.exposeAsEntityConfig) !== null && _a !== void 0 ? _a : '',
            };
            newSchema.exposeToHomeAssistant = undefined;
            newSchema.haConfig = undefined;
            return newSchema;
        },
    },
];
